cordova-plugin-outline/android/java/org/outline/OutlinePlugin.java

```diff
@@ -284,135 +282,104 @@ public void onActivityResult(int request, int result, Intent data) {
       LOG.warning("Received non-requested activity result.");
       return;
     }
-    if (result != RESULT_OK) {
-      LOG.severe("Failed to prepare VPN.");
-      PluginResult pluginResult =
-          new PluginResult(PluginResult.Status.ERROR, ErrorCode.VPN_PERMISSION_NOT_GRANTED.value);
-      sendPluginResult(startRequestTunnelId, Action.START.value, pluginResult, false);
-      return;
-    }
-    try {
-      startVpnTunnel();
-    } catch (Exception e) {
-      // Do not propagate the exception in the main thread; the service will broadcast the error.
+    if (result != Activity.RESULT_OK) {
+      LOG.warning("Failed to prepare VPN.");
     }
+    vpnPreparedSignal.countDown();
   }
 
-  private void startVpnTunnel() {
+  private int startVpnTunnel(final String tunnelId, final JSONObject config) {
     LOG.info("Starting VPN Tunnel");
-    if (vpnTunnelService == null) {
-      onVpnTunnelServiceNotBound(Action.START, startRequestTunnelId);
-      return;
+    final TunnelConfig tunnelConfig = new TunnelConfig();
+    tunnelConfig.id = tunnelId;
+    tunnelConfig.proxy = new ShadowsocksConfig();
+    try {
+      tunnelConfig.name = config.getString("name");
+      tunnelConfig.proxy.host = config.getString("host");
+      tunnelConfig.proxy.port = config.getInt("port");
+      tunnelConfig.proxy.password = config.getString("password");
+      tunnelConfig.proxy.method = config.getString("method");
+    } catch (JSONException e) {
+      LOG.log(Level.SEVERE, "Failed to retrieve the tunnel proxy config.", e);
+      return ErrorCode.ILLEGAL_SERVER_CONFIGURATION.value;
+    }
+    try {
+      return vpnTunnelService.startTunnel(tunnelConfig);
+    } catch (Exception e) {
+      LOG.log(Level.SEVERE, "VPN service not bound", e);
+      return ErrorCode.UNEXPECTED.value;
     }
-    vpnTunnelService.startTunnel(startRequestTunnelId, startRequestConfig);
   }
 
-  private void stopVpnTunnel(final String tunnelId) {
+  private int stopVpnTunnel(final String tunnelId) {
     LOG.info("Stopping VPN tunnel.");
-    if (vpnTunnelService == null) {
-      onVpnTunnelServiceNotBound(Action.STOP, tunnelId);
-      return;
+    try {
+      return vpnTunnelService.stopTunnel(tunnelId);
+    } catch (Exception e) {
```

Comment:
Given the try-catch in `executeAsync`, is this one necessary?

Reply:
Not really, I thought it would make logging more readable. I have improved logging in `executeAsync` and removed this method.